### PR TITLE
Proper error handling for basic auth

### PR DESCRIPTION
**Pull request by Kayvan-Ehtesh**

### CODE REVIEW

File: class-auth.php [hunk 1]
```diff
@@ -224,7 +224,7 @@ public function generate_token( $user, $return_raw = true ) {
 			'exp'  => $expire,
 			'data' => array(
 				'user' => array(
-					'id' => $user->ID,
+					'id' => $user->id,
```
Comment (btilford) [inline class-auth.php]: Was this case change intentional? WP tends to use upper on `ID`.

File: class-auth.php [hunk 2]
```diff
@@ -321,11 +321,15 @@ public function validate_token( $return_response = true ) {
...
 		}
 
-		$hash = end(explode("_", AUTH_COOKIE));
+        $auth_arr = explode("_", AUTH_COOKIE);
+
+		$hash = end($auth_arr);
 
+		//Check if basic auth is also present
 		if(! $auth) {
 			foreach(array_keys($_COOKIE) as $key) {
```
Comment (btilford) [inline class-auth.php]: Is it possible to access the auth token using key index access instead of a loop?

```
$token  = $_COOKIE[$hash]; // not sure what $hash is assuming it's the cookie name for the token?

if($token) ...
```

File: class-auth.php [hunk 3]
```diff
@@ -351,14 +355,15 @@ public function validate_token( $return_response = true ) {
...
 		if (!$token) {
-            // Check if Basic Auth is used and create a token if so
-            list($username, $password) = explode( ':', base64_decode( substr( $auth, 6 ) ) );
-            $request = new WP_REST_Request( 'POST', '/wp-json/jwt-auth/v1/token' );
-            $request->set_param( 'username', $username );
-            $request->set_param( 'password', $password );
-            $token = $this->generate_token( $request );
-            if (is_array($token) && isset($token['token'])) $token = $token['token'];
-            return;
+            return new WP_REST_RESPONSE(
```
Comment (btilford) [inline class-auth.php]: This is sent if the token **doens't ** exist?

File: class-auth.php [hunk 4]
```diff
@@ -351,14 +355,15 @@ public function validate_token( $return_response = true ) {
...
+                    'success'       => true,
+                    'statusCode'    => 200,
+                    'code'          => 'basic_auth_valid',
+                    'message'       => __('Basic auth is valid', 'jwt-auth'),
+                    'data'          => array()
+                )
+            );
         }
 
 		if ( ! $token ) {
```
Comment (btilford) [inline class-auth.php]: Hard to follow in the PR column but didn't we just send a response for `!$token` ? Maybe the JWT stuff can be encapsulated in it's own function if we are nesting too far here.

File: class-auth.php [hunk 5]
```diff
@@ -351,14 +355,15 @@ public function validate_token( $return_response = true ) {
...
-            return;
+            return new WP_REST_RESPONSE(
+                array(
+                    'success'       => true,
+                    'statusCode'    => 200,
+                    'code'          => 'basic_auth_valid',
+                    'message'       => __('Basic auth is valid', 'jwt-auth'),
+                    'data'          => array()
+                )
+            );
```
Comment (btilford) [inline class-auth.php]: Is anything actually parsing and verifying the token?

**Comment by btilford** [comment]
@Kayvan-Ehtesh had a few questions, I think there is likely code from the last version I'm just not seeing in the PR most likely.

File: class-auth.php (hunk 3 above)
Comment (Kayvan-Ehtesh) [reply class-auth.php]: Yes, so the `!auth` logic above is checking for basic authentication, if there is no basic authentication and no token it returns errors, if there is basic authentication but no token we still want to return a success because the wp-admin panel doesn't have jwts it just uses basic auth. This return statement is to allow things like Manage Partners to keep working without jwt tokens

File: class-auth.php (hunk 4 above)
Comment (Kayvan-Ehtesh) [reply class-auth.php]: Thanks for catching this, didn't realize there was a duplicate if

File: class-auth.php (hunk 5 above)
Comment (Kayvan-Ehtesh) [reply class-auth.php]: Not here, since this is logic for authentication without a token, like in the wp-admin panel. Further down, probably not in view of the PR there is

File: class-auth.php (hunk 2 above)
Comment (Kayvan-Ehtesh) [reply class-auth.php]: I attempted that when we first setup the plugin, I think I was running into issues involving something inconsistent while we were trying to automate grabbing a jwt token for our endpoint check

I can create a ticket and dig into it further

Comment (Kayvan-Ehtesh) [reply class-auth.php]: Looked back over it, it was that the index could be multiple values depending on where it was used, 'nonce' 'X-WP-Nonce' etc.

Comment (btilford) [reply class-auth.php]: Main thing is that this code may run every request so even with a limit being on cookie size it is an unnecessary loop.

File: class-auth.php (hunk 3 above)
Comment (btilford) [reply class-auth.php]: Cool, might make sense to add a comment that we are accepting the basic auth or something like that.

Another pattern that helps with readability is single return  per function _(combined with fail fast or exceptional conditions evaluated first)_, this especially helps in statically typed languages or if there is a linter because those things can pick up any path that doesn't return.

https://coderanch.com/t/685748/java/Single-return

Comment (Kayvan-Ehtesh) [reply class-auth.php]: I think some of the issue with single return per function here is that the repo I forked it from has a lot of validation that is going to fail after this if statement because it doesn't have a jwt token. I will make it more readable with comments but I think I'd need to rewrite a few things to make single return workable for this function.

**Review by btilford — APPROVED** [review]

**Review by btilford — APPROVED** [review]
Looks good, shouldn't really affect the basic/default auth anyways so seems pretty safe.
